### PR TITLE
What's new and readme: we are moving to Python 3.7

### LINKED ISSUE #9862: Update changes file on threshold branch
Look at the PR's merged to `threshold` and `threshold_py3_staging`, decide if a what's new entry is required and write one if so. Particular attention should be given to changes relevant to developers (especially authors of, add-ons, braille drivers, synth drivers). Non user visible changes do not need to be included.

Please list the PR's that you have looked at (and either written an entry, or decided one isn't necessary) in the description, all PR's to `threshold` or `threshold_py3_staging` need consideration. 

Because this will be on-going work create a draft PR, it helps us to know when the PR is finished, or more accurately, that it is not. We likely wont merge it until we are ready to merge threshold back to master. Because only then we will know we have handled all PR's

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
- Fixes #9862 
Gives closure to #7105 

### Summary of the issue:
What's new and readme still mentions Python 2.7, and in case of what's new, changes were needed.

### Description of how this pull request fixes the issue:
Edits include mentioning Python 3.7 in both readme and what's new, as well as fixing things here and there in what's new/Threshold section. See the below section regarding work included in what's new and readme through this pull request and other Threshold work.

### Testing performed:
None

### Known issues with pull request:
Some parts are missing, such as filling in which issue introduced support for pyserial 3.4 and other things. What's new and readme are subject to change based on comments.

### Change log entry:
None

### Pull requests addressed by this work

Items with a checkmark are mentioned in what's new. Merge messages are ignored.

#### Speech refactor

- [x] New speech framework including callbacks, beeps, sounds, profile switches and prioritized queuing (#7599)

#### Python 3.7

- [x] nvwave.playWaveFile: async > asynchronous (#8647)
- [x] Update Developer Guide re. tab completion (PR #9935)
- [x] Update to pySerial 3.4 (#8815)
- [x] Remove functionality marked as deprecated (#9603)
- [x] Update dependencies for Python 3 (#9630)
- [x] Bump NVDA version for threshold to 2019.3.0 and also set backwards compat to the same. We already know there will be 2019.2 before threshold, but this may need to be bumpt to 2019.4.0 if we end up doing a 2019.3 off master as well.
- [x] Compile liblouis with 32 bit widechars (#9544)
- [x] textUtils module to deal with offset differences between Python 3 strings and Windows wide character strings with surrogate characters (PR #9545)
- [x] [No entry necessary] - Python 3: Update py2exe to fix several issues (PR #9856)
- [x] Update UIAutomationCore.dll comtypes module to version 10.0.18362 (PR #9829)
- [x] Bump threshold version and add-on compatibility version to 2019.4.0 for now as master is already heading toward a 2019.3.
- [x] No longer require pyWin32 (#9639)
- [x] Appveyor: use Python 3.7 rather than Python 2.

## No entry necessary
- [x] Fixes for BrailliantB and FreedomScientific (PR #9839)
- [x] Python 2/3: relative imports (#9670)
- [x] Python 3: print statement -> print function (#9664)
- [x] Python 3: file function -> open function (#9663)
- [x] Python 3: use floor division (//) to obtain integer results from division operations (#9692)
- [x] Python 3: use importlib for importing modules (#9693)
- [x] Python 3: dictionary iteration/list methods made compatible with Python 3 (#9671)
- [x] Python 3: xrange -> range (#9627)
- [x] SAPI 5 synth driver: percentage to pitch/rate must return an integer. Re #9641. (#9699)
- [x] Python 3: except exception, e -> except as e (#9698)
- [x] Python 3: dict.haskey -> key in dict (#9697)
- [x] Check pot: file -> open (#9718)
- [x] WinConsole handler/Python 3: use floor divisoin assignment operator. Re #9641. (#9737)
- [x] Unittests/Python 3: update syntax, catch unbound method registration test in extension points as expected failure (#9726)
- [x] Python 3: zip function -> list(zip), itertools.izip -> zip() (#9729)
- [x] Global commands: restore commands affected by division operator differences between Python 2 and 3 (#9748)
- [x] Python 3: more native imports and modules (#9751)
- [x] Input core/input gestures dialog: script function object has no attribute __func__. (#9752)
- [x] Python 3: use native metaclass definition syntax (#9753)
- [x] System tests/Python 3: syntax and imports (#9801)
- [x] App module handler/does app mod exists: return false when called before imports are initialized  (#9800)
- [x] Add-on handler: ugettext -> gettext (#9810)
- [x] Python 3/urllib: imports, timeout (#9820)
- [x] Update check/Python 3: update check resource is encoded as bytes, thus make it Unicode. Re #9819. (#9837)
- [x] WinKernel: create a wrapper for MoveFileExW function and associated flag constants (PR #9849)
- [x] Python 3: use UTF-8 decoding for font name with Scintilla controls (PR #9893)
- [x] Python 3:  MoveFileEx - catch WindowsError instead of WinError (PR #9876)
- [x] Python 3: Use built-in "input" instread of "raw_input" (PR #9908)
- [x] Python 3: Check for None during check for previous installation version in Installer GUI (PR #9924)
- [x] Update setup script for Python 3.7 (#9605)
- [x] Update Scons build environment to run on Python 3 (#9667)
- [x] Python 3: Fix use of subprocess (#9716)
- [x] Python 3: No longer encode or decode if it is not necessary (#9734)
- [x] Python 3: fix several remaining division errors (#9774)
- [x] Write syntax errors and tracebacks to python console (#9767)
- [x] Python 3: fix logging level not being updated correctly (#9809)
- [x] Python 3: Fix deprecation warning extensionPoints.callWithSUpportedKwargs (PR #9778)
- [x] Fix comparison error in start menu (PR #9846)
- [x] Update BRLTTY driver (PR #9821)
- [x] Python 3 one core: Don't check for sys.frozen in the check method. (PR #9857)
- [x] Python 3: fix say all for DisplayModelEditableText (PR #9866)
- [x] Python 3: sys.exec_prefix > sys.prefix for nvda_slave (PR #9871)
- [x] Python 3: Fix usages of "open" function (PR #9853)
- [x] Python3: Fix save scripts in lockapp (PR #9884)
- [x] Python 3: When processing xml coming from virtual buffers, make sure that surrogate characters are handled properly (PR #9897)
- [x] Python 3: Remove many instances of "py3 review required" comments (PR #9890)
- [x] Python 3: Fix adobe reader int/str to NoneType comparisons as well as other identified cases (PR: #9881)
- [x] Revert "Restore VoiceSettingsSlider to ensure we don't break compatibility with Audio Themes add-on" (PR #9907)
- [x] Python 3: Clean the current words buffer in more cases (PR #9877)
- [x] Python 3: Remove Python 2 specific workarounds (PR #9889)
- [x] Python 3: Fix loading of espeak variants (PR #9920)
- [x] Python 3: Fix several local overrides of range function (PR #9912)
- [x] Fix waiting for outlook/powerpoint window to exist for ever (PR #9950)
- [x] Log never handled speech indexes, and don't make the speech queue lag behind (PR #9946)
- [x] Fix error in JAVA text ranges (PR #9951)
- [x] appveyor: build snapshots for threshold
- [x] OneCore synthDriver: fix up merge conflicts.
- [x] Only clear curWordChars on focus / caret movement (#9597)
- [x] import Python 3 native modules (#9703)
- [x] Python 3: sys.maxint -> sys.maxsize, as maxint is gone in Python 2. Re #7105. (#9704)
- [x] Python3: os.getcwdu -> os.getcwd (#9705)
- [x] Python 3/remote python console: SocketServer -> socketserver. (#9706)
- [x] Update miscDeps to latest master, containing python3 txt2tags.
- [x] Python 3: the thread module is no longer available. Use threading instead. (#9713)
- [x] Python3: fix remaining syntax errors found by py2exe (#9719)
- [x] Python3: fix NVDA's logging infrastructure (#9712)
- [x] Python3: gettext is unicode, and pgettext is placed in builtins properly (#9714)
- [x] objidl: HRESULT is no longer available in ctypes.wintypes. Get it from comtypes. (#9745)
- [x] remove use of unichr, basestring and unicode objects. (#9724)
- [x] Remove need for hashing NVDAObject and TextInfo objects (#9746)
- [x] Revert "Remove need for hashing NVDAObject and TextInfo objects (#9746)" (#9756)
- [x] Python3: base virtualBuffer: coerce attribute values to a string before escaping for regular expression  (#9759)
- [x] cmp no longer exists in Python3. Replace with (a>b)-(a<b) (#9758)
- [x] Python3: Provide __hash__ methods where ever we have provided __eq__ methods (#9757)
- [x] Correct a few places where we are trying to do integer comparisons with None. (#9794)
- [x] Ensure NVDA contains a signed and valid python37.dll (#9796)
- [x] Python3: implement new way of converting COMError to CallCancelled (#9795)
- [x] bdDetect: change basestring to str. This was the only thing causing NVDA not to run in this branch, and will be introduced exactly the same by pr #9736 anyway.
- [x] Make output of both py2exe and nose more verbose to aide in future build issues.
- [x] Revert "App module handler/does app mod exists: return false when called before imports are initialized  (#9800)" (#9815)
- [x] config.addConfigDirsToPythonPackagePath: When adding scratchpad dirs, create them if they don't already exist. (#9814)
- [x] Update miscDeps to contain new brlApi module (#9812)
- [x] Python3: stop using unichr in brailleInput.py (#9835)
- [x] Python3 convert braille drivers (PR #9736)
- [x] Fix buildversion for threshold branch
- [x] Fix api back compat version number
- [x] Python 3: Fix missing import, remove unused imports (PR #9901)
- [x] Python 3: Fix error with accPropServer (PR #9900)
- [x] txt2tags for python3 #8734 (#9648)
- [x] Change iter.next() to next(iter) (PR #9535)
- [x] Remove .pyc and .pyo files when updating (PR #9960)
- [x] Python3: Fix SAPI5 and MSSP on installed NVDA (PR #9964)
- [x] Detect caret movement using events, except for IA2Web #9933

### CODE REVIEW

**Comment by josephsl** [comment]
Hi all,

Comments from everyone who has worked on or commented on Python 3 are appreciated, especially from Reef, @JCSTeh, and @leonardder.

Thanks,

File: user_docs/en/changes.t2t [hunk 1]
```diff
@@ -10,12 +10,12 @@ What's New in NVDA
...
-- Updated pySerial to version 3.4
+- Updated pySerial to version 3.4.
 - the validate module is now only available from configobj. Code should now do from configobj import validate rather than import validate.
-- textInfos.Point and textInfos.Rect are removed and replaced by locationHelper.Point and locationHelper.RectLTRB, respectively 
+- textInfos.Point and textInfos.Rect are removed and replaced by locationHelper.Point and locationHelper.RectLTRB, respectively.
 - braille.BrailleHandler._get/set_tether have been removed. use getTether and setTether respectively.
-- config.getConfigDirs has been removed
-- config.ConfigManager.getConfigValidationParameter is removed. A bug in getConfigValidation has been fixed in order for it being able to fully replace getConfigValidationParameter
+- config.getConfigDirs has been removed.
+- config.ConfigManager.getConfigValidationParameter is removed. A bug in getConfigValidation has been fixed in order for it being able to fully replace getConfigValidationParameter.
```
Comment (XLTechie) [inline user_docs/en/changes.t2t]: Can this line possibly be made more clear? Something like this, perhaps?

```suggestion
- config.ConfigManager.getConfigValidationParameter has been removed. A bug in getConfigValidation was fixed in order for it to be able to fully replace getConfigValidationParameter.
```

**Comment by josephsl** [comment]
Hi, right, thanks for catching this, which also had a grammar problem (will resolve it once I wake up from sleep in the morning). Thanks.

File: readme.md [hunk 2]
```diff
@@ -35,7 +35,7 @@ The NVDA source depends on several other packages to run correctly.
 ### Installed Dependencies
 The following dependencies need to be installed on your system:
 
-* [Python](http://www.python.org/), version 2.7.15, 32 bit
+* [Python](http://www.python.org/), version 3.7.4, 32 bit
```
Comment (leonardder) [inline readme.md]: Please consider #8421

Note that on appveyor, we can't ensure that we are on a particular minor release. Therefore, I propose that we advice to always use the most recent minor version

```suggestion
* [Python](http://www.python.org/), most recent minor release of version 3.7, 32 bit
```

File: user_docs/en/changes.t2t [hunk 3]
```diff
@@ -10,12 +10,13 @@ What's New in NVDA
 
 
 == Changes for Developers ==
- - Updated pySerial to version 3.4
+- Updated Python to 3.7.4. (#7105)
```
Comment (leonardder) [inline user_docs/en/changes.t2t]: Same as above.

File: user_docs/en/changes.t2t [hunk 4]
```diff
@@ -10,12 +10,13 @@ What's New in NVDA
 
 
 == Changes for Developers ==
- - Updated pySerial to version 3.4
+- Updated Python to 3.7.4. (#7105)
+- Updated pySerial to version 3.4.
 - the validate module is now only available from configobj. Code should now do from configobj import validate rather than import validate.
-- textInfos.Point and textInfos.Rect are removed and replaced by locationHelper.Point and locationHelper.RectLTRB, respectively 
+- textInfos.Point and textInfos.Rect are removed and replaced by locationHelper.Point and locationHelper.RectLTRB, respectively.
 - braille.BrailleHandler._get/set_tether have been removed. use getTether and setTether respectively.
```
Comment (leonardder) [inline user_docs/en/changes.t2t]:
```suggestion
- braille.BrailleHandler._get/set_tether have been removed. Use getTether and setTether respectively.
```

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
You can change the threshold release heading to 2019.3

File: user_docs/en/changes.t2t (hunk 1 above)
Comment (codeofdusk) [reply user_docs/en/changes.t2t]: Even better, you can remove the words "in order" from that sentence.

**Comment by josephsl** [comment]
Hi, I knew there was an issue about mentioning minor versions in docs – thanks for the link to the issue. Thanks.

**Comment by josephsl** [comment]
Hi,

Should we mention textUtils?

Thanks.

**Comment by leonardder** [comment]
yes, sounds like a good idea. Probably best to point at examples from the changelog.

**Comment by josephsl** [comment]
Hi,

Regarding text utils, I have nothing to say - can't come up with a descriptive changelog for developers at the moment.

Anything else from our PR pool? If not, I'll switch the PR to review mode.

Thanks.

**Comment by leonardder** [comment]
How about something within the lines of:

* Added a new textUtils module that makes it easier to deal with string differences between Python 3 strings and Windows unicode strings. See the module documentation and textInfos.offsets module for example implementations. (#9545)

**Comment by josephsl** [comment]
Hi,

Text utils: done.

Thanks.

**Review by feerrenrut — APPROVED** [review]
This is looking pretty good. Please update the description to cover which PR's (to threshold / threshold staging) have been covered by this. You can get most of the way there by copying entries from git log on the threshold branch. I suggest using the checkbox markdown:

```
- [ ] some PR not covered
- [x] some PR covered
```

**Comment by josephsl** [comment]
Hi,

I'll use git shortlog to edit the description.

Thanks.

**Comment by josephsl** [comment]
Hi,

To @LeonardDer: whenever you are ready, please take a look at this PR. Thanks.

**Comment by feerrenrut** [comment]
Thanks @josephsl, I have updated the description so that they render as a checkboxes. It looks like there are still many PR's that need consideration.

**Comment by josephsl** [comment]
Hi, yes, including the just merged one (#9960). Thanks.

**Comment by feerrenrut** [comment]
If no one volunteers before tomorrow (25-July 0600 UTC+0) , then I'll finish this off.

File: user_docs/en/changes.t2t [hunk 5]
```diff
@@ -3,31 +3,33 @@ What's New in NVDA
...
- - Updated pySerial to version 3.4
+- Updated Python to 3.7. (#7105)
+- Updated pySerial to version 3.4.
 - the validate module is now only available from configobj. Code should now do from configobj import validate rather than import validate.
-- textInfos.Point and textInfos.Rect are removed and replaced by locationHelper.Point and locationHelper.RectLTRB, respectively 
-- braille.BrailleHandler._get/set_tether have been removed. use getTether and setTether respectively.
-- config.getConfigDirs has been removed
-- config.ConfigManager.getConfigValidationParameter is removed. A bug in getConfigValidation has been fixed in order for it being able to fully replace getConfigValidationParameter
-- inputCore.InputGesture.logIdentifier property has been removed. Now use inputCore.InputGesture.identifiers[0] instead.
+- textInfos.Point and textInfos.Rect has been removed and replaced by locationHelper.Point and locationHelper.RectLTRB, respectively.
```
Comment (leonardder) [inline user_docs/en/changes.t2t]:
```suggestion
- textInfos.Point and textInfos.Rect have been removed and replaced by locationHelper.Point and locationHelper.RectLTRB, respectively.
```

File: user_docs/en/changes.t2t [hunk 6]
```diff
@@ -3,31 +3,33 @@ What's New in NVDA
...
+- braille.BrailleHandler._get/set_tether has been removed. Use getTether and setTether respectively.
```
Comment (leonardder) [inline user_docs/en/changes.t2t]:
```suggestion
- braille.BrailleHandler._get/set_tether have been removed. Use getTether and setTether respectively.
```

**Review by LeonarddeR — DISMISSED** [review]
I found two sentences where has should be have IMO. Could be wrong for the latter, though.

**Comment by josephsl** [comment]
Hi, I think part of that may have to do with subtle English grammar differences.

**Comment by josephsl** [comment]
Hi,

Let me at least commit the requested edits, and then I think we should decide our next steps.

Thanks.

**Comment by feerrenrut** [comment]
I'd like to get this finished off and merged in, are you going to continue with it @josephsl? I don't think I will hold back threshold any longer.

**Comment by josephsl** [comment]
Hi, feel free to merge – let’s get the party started. Thanks.

**Comment by josephsl** [comment]
Hi,

Clarification: I'm done with this one.

Thanks.

**Comment by leonardder** [comment]
@josephsl: could you please do a base branch switch for this?

**Comment by josephsl** [comment]
Done.

**Comment by derekriemer** [comment]
Should this be merged?

**Comment by josephsl** [comment]
Hi, when it is ready, yes. Thanks.

**Comment by feerrenrut** [comment]
@derekriemer The checklist for PR's needs to be completed, I suspect many of these do not need an entry. This is on my list of things to do, but hasn't yet received priority.

**Comment by feerrenrut** [comment]
I have gone through the unchecked items and added entries.

I also re-arranged the current items in the changes file.

**Pull request merged**